src/dotnet-core-uninstall/Shared/VSVersioning/VisualStudioSafeVersionsExtractor.cs

```diff
@@ -61,48 +61,52 @@ private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) A
                 .Select(pair => (pair as IEnumerable<Bundle>, LocalizableStrings.MacRuntimeRequirementExplanationString))
                 .ToDictionary(key => key.Item1, value => value.Item2); 
 
-            var sdks = bundleList.Where(bundle => bundle.Version is SdkVersion);
-            if (sdks != null && sdks.Count() > 0)
+            if (macOSPreserveVSSdks)
             {
-                dividedBundles.Add(sdks, LocalizableStrings.MacSDKRequirementExplanationString);
+                var sdks = bundleList.Where(bundle => bundle.Version is SdkVersion) ?? [];
+                if (sdks.Any())
+                {
+                    dividedBundles.Add(sdks, LocalizableStrings.MacSDKRequirementExplanationString);
+                }
             }
 
             var remainingBundles = bundleList
-                .Where(bundle => !(bundle.Version is RuntimeVersion || bundle.Version is SdkVersion))
+                .Where(bundle => bundle.Version is not RuntimeVersion && !(bundle.Version is SdkVersion && macOSPreserveVSSdks))
                 .Concat(bundlesAboveLimit);
             return (dividedBundles, remainingBundles);
         }
 
-        private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) ApplyVersionDivisions(IEnumerable<Bundle> bundles)
+        private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) ApplyVersionDivisions(IEnumerable<Bundle> bundles, bool macOSPreserveVSSdks = false)
         {
             if (RuntimeInfo.RunningOnWindows)
             {
                 return ApplyWindowsVersionDivisions(bundles);
             }
             else
             {
-                return ApplyMacVersionDivisions(bundles);
+                return ApplyMacVersionDivisions(bundles, macOSPreserveVSSdks);
```

Comment:
Is the first divided bundles important? I'd expected to just condition having this call at all on macOSPreserveVSSdks

Reply:
Yes. From what I understand, while the sdks are all uninstallable now, runtimes should still be checked to avoid uninstalling a runtime associated to an sdk

Reply:
If you uninstall an SDK, why do you need the runtime associated with that SDK?

Reply:
I thought it more like the other way around, when people attempt to uninstall a runtime associated with an sdk

Reply:
So then you mean it should not compare runtimes and sdks when uninstalling either?

Reply:
So like let's say you have SDKs sa, sb, sc, and sd installed right now, and you have runtime ra, rb, and rc. sa depends on ra, sb on rb, and both sc and sd depend on rc. If you're planning to uninstall sa and sc and keep sb and sd, then you should uninstall ra because no SDK that isn't being uninstalled requires it. You can't uninstall rb because you aren't uninstalling sb, and you can't install rc because sd requires it even though you're uninstalling sc. Is that clearer?

Reply:
Yes, thanks. So in essence, when uninstalling an sdk, it should also remove unused runtimes, right?

Reply:
Yeah, for mac I think that's right.

Windows get a little more confusing because we don't really look for VS instances properly, so we (conservatively) didn't uninstall runtimes VS might need. We used to do the same for mac, but since VS for mac isn't really a thing anymore, we don't have to.